li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaInstrumentedConsumerImpl.java

```diff
@@ -776,6 +744,9 @@ private boolean proceedClosing() {
     if (isClosed()) {
       return false;
     }
+    closing = true;
```

Comment:
I'm not entirely sure this is correct, especially since proceedClosing() can be called from Close() which doesn't acquire a lock. I think it would be safest to replace this `volatile boolean` with an AtomicBoolean and use [AtomicBoolean#compareAndSet(boolean, boolean)](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/atomic/AtomicBoolean.html#compareAndSet(boolean,%20boolean)) to guarantee that you don't have a situation where two threads race to set `closing = true`.

Reply:
I at least think it would simplify the logic in proceedClosing a bit

Reply:
Just as an example problem, you could have two threads setting close, assigning closer, and then try to acruire delegateLock. Only one will get to that point, but the code that acquired delegateLock first *might not* be the last one to set `closer`.

So I think this is better:

```
if (!closing.compareAndSet(false, true)) {
  return;
}
```

Or however you use that correctly.

Reply:
i've reworked the patch, and addressed these issues (i think)